### ISSUE TITLE
Model should allow multiple Selectors per SpecificResource

### ISSUE THREAD

**Issue by azaroth42** [w3c/web-annotation#207]
In https://www.w3.org/TR/annotation-model/#selectors, we say that there MUST be 0..1 selectors.

I think this is a bug, given the decision in https://github.com/w3c/web-annotation/issues/93#issuecomment-186321067 and it should be 0..Many.  Note that all of the other specifiers (State, Scope, Style, Purpose, Rendering) are already 0..Many.

**Comment by iherman** [comment]
This is closely related to issue #205 : what does multiple selector mean. Although #93, I believe, was mainly about the exact syntax to be adopted for a refinement, the comment indeed suggests that we may have several selectors and they would represent a choice. Ie, related to my comment in #205 (https://github.com/w3c/web-annotation/issues/205#issuecomment-213665437), this seems to indicate a disjunction.

Fine... but you/we should explain what that means. Does it mean that the client can choose among two selectors? Following what rules? Or is it entirely in the hands of the implementation? How do we have interoperability? Do we expect that both selectors select, at the end of the day, the same portion of a resource, or is it allowed to have different ones? What are the use cases?

B.t.w., I believe that, in fact, this issue is the same as #205. Whatever we decide here should also be the guiding argument to solve #205 and vica versa...

**Comment by azaroth42** [comment]
Right :) That's how I came to the (IMO) bug. 

My understanding from #93 is:
- If there are multiple States/Selectors/Styles/Renderings/... then you pick one (previously it was an explicit oa:Choice)
- If you need to have more than one and they all get processed in order, that's where `refinedBy` comes in. (previously it was an explicit oa:List)
- If you need more than one and the order doesn't matter ... then just put them in some order using `refinedBy`. (previously an explicit oa:Composite)

The docs for refinedBy say: 

> A Selector may be refinedBy 1 or more other Selectors. If more than 1 is given, then they are considered to be alternatives that will result in the same selection.

And my proposal is that this should be copied to `selector` as well (with clarification that they may not all have the same fidelity)

**Comment by azaroth42** [comment]
> But... doesn't that leads to interoperability issues. You pick, randomly, one and I pick the other? What is the use case for this?

The idea is that the more you specify, the more likely it is that one will continue to work and that clever clients can use the set to be even more precise.  For example, if you use just TextPosition, you have no way to tell that the document changed, but it's as precise a way as possible when the document is static.  So if there's also a TextQuote selector, then you can use that to either find the selection (even though it's maybe slower or harder), or to verify that the content at the specified position is the same. 
This is towards robustness of anchors, but clearly not the entire solution.

Second use case is if one system implements only TextQuote, then you can use the annotation, whereas if you supplied just a complex RangeSelector of XPathSelectors, that might be more accurate and faster ... but worthless to the simpler implementation of just the Quote pattern.  Related to this is if there are new, super awesome selectors that are developed in different communities, it would be good to at least provide one of the selectors from the model for systems outside that community.

**Comment by iherman** [comment]
@azaroth42:

> > But... doesn't that leads to interoperability issues. You pick, randomly, one and I pick the other? What is the use case for this?
> 
> The idea is that the more you specify, the more likely it is that one will continue to work and that clever clients can use the set to be even more precise. For example, if you use just TextPosition, you have no way to tell that the document changed, but it's as precise a way as possible when the document is static. So if there's also a TextQuote selector, then you can use that to either find the selection (even though it's maybe slower or harder), or to verify that the content at the specified position is the same.
> This is towards robustness of anchors, but clearly not the entire solution.
> 
> Second use case is if one system implements only TextQuote, then you can use the annotation, whereas if you supplied just a complex RangeSelector of XPathSelectors, that might be more accurate and faster ... but worthless to the simpler implementation of just the Quote pattern. Related to this is if there are new, super awesome selectors that are developed in different communities, it would be good to at least provide one of the selectors from the model for systems outside that community.

Ok, these are all good issues. I believe the underlying thought is that if there are two selectors (or a state and a selector, reflecting to issue #205? Or is that a different issue after all?), they **_SHOULD (MUST?) select the same content**_.

My problem is how to spec this. Is this a testable statement for an implementation? How would an implementation be sure about this? How to ensure interoperability? Or should we just stop by saying 'MUST' and leave the rest to implementations?

**Comment by tbdinesh** [comment]
I guess leaving it at SHOULD makes sense.

Seeing multiple selection as not-options but as alternates is good example.
And in the presence of state, we can even think of them as options as a use
case.

Imagining a scenario:
say we have linguistic states
such as en-state, nl-state, .. kn-state

then the alternates would be specific to the language
with say the quoted-text options being "one must", "een moet", ..

**Comment by iherman** [comment]
@tbdinesh :

> Imagining a scenario:
> say we have linguistic states
> such as en-state, nl-state, .. kn-state
> 
> then the alternates would be specific to the language
> with say the quoted-text options being "one must", "een moet", ..

Well... this example actually shows the ambivalence of the situation: that there is no way to check whether a multi-selector usage is correct or not.

Indeed:
- If the goal is to annotate a particular, say, quote in a book in the sense of quoting from the "work" (say, I want to annotate the first chapter of a particular novel) then yes, this example is fine because I offer the alternative for various languages.
- However, if I want to annotate the translations of a book, ie, I want to make a comment on the Dutch but also the Kannada translations, I would think that the example is incorrect, because I should have a separate annotation (ie, separate targets with Specific Resources) on the two translations, and not bound into one with a multiple selector.

There is a fine line between these two, and it may be a kind of a judgement call on whether a particular set up is acceptable or not. Put it another way, whether the two (or more) selectors refer to the same content is not an independently verifiable claim...

Maybe the correct approach is a MAY (as in "multiple selectors MAY be used if they refer to the same target"), with a note making it clear that the usage of that particular selection may be different from one application to the other and may lead to interoperability issues. I can live with that.

**Comment by tbdinesh** [comment]
I see your point, and agree for MAY.

I was trying bring out the diversity of interpretations of "state".
An html-state versus a pdf-state is not very different than an en-state
versus a kn-state
except maybe for quotable text possibilities - where we assume certain text
extractors. Where
as using the linguistic states example was only to simplify the selector
usage in the example, although foreseeable in renarration services. In
anycase, it was suggesting that if we assume verifiability, we will only be
delegating these ambivalences to another level where it is not as
transparent.

**Comment by azaroth42** [comment]
I'd like to argue that multiple selectors SHOULD select the same content (with sufficient flexibility to distinguish intellectual content in different languages, versus the same exact string in different formats/representations), and user agents SHOULD anticipate that some selectors will not have as fine grained selection capabilities as others. The UA MUST pick one of the selected segments, if different segments are described.  
[Multiple targets is just that... multiple _targets_]

**Comment by pciccarese** [comment]
Assume I want to target  two non contiguous text fragments as a single unit and the annotation does not make sense for one of them taking singularly. Example, I want to annotate the full name (Paolo+Ciccarese) in: "the name Paolo turned out to be of a man belonging to the Ciccarese's family". 

I cannot express such annotation precisely with multiple targets ("it is also possible for an Annotation to have multiple Bodies and/or Targets. In this case, each Body is considered to be equally related to each Target individually, rather than to the set of all of the Targets."). 

It is not a Choice, which current Example Use Case (3.3.1) is about languages. 

As I understand, I also could not use multiple selectors as they are assumed to select the same content. 

Refinement seems a bit overkill as the two fragments could be very far apart in the content.

What are the options I am left with aside of being vague?

In other words how do we clearly represent the three flavors:
- multiple targets and independent
- multiple targets as single units
- alternative targets

**Comment by iherman** [comment]
@azaroth42 : you propose

> multiple selectors SHOULD select the same content (with sufficient flexibility to distinguish intellectual content in different languages, versus the same exact string in different formats/representations), and user agents SHOULD anticipate that some selectors will not have as fine grained selection capabilities as others. The UA MUST pick one of the selected segments, if different segments are described.

I guess what it means, referring back to my [earlier comment](https://github.com/w3c/web-annotation/issues/207#issuecomment-216022290) and @tbdinesh 's example, this means that

> However, if I want to annotate the translations of a book, ie, I want to make a comment on the Dutch but also the Kannada translations, I would think that the example is incorrect, because I should have a separate annotation (ie, separate targets with Specific Resources) on the two translations, and not bound into one with a multiple selector.

is indeed incorrect, and the author should use multiple targets itself.

All things considered, I am fine with it. I would propose, however, to an an (informative, obviously) note to the text which makes it clear that authors should use the possibility for a multiple target, when appropriate, to avoid the semantic pitfalls.

(Note that the resolution on this issue should also be valid for issue #205)

**Comment by iherman** [comment]
@paolociccarese 

> In other words how do we clearly represent the three flavors:
> - multiple targets and independent
> - multiple targets as single units
> - alternative targets

While I understand this is a concern, I would be a big worried about introducing a standard mechanism to define these alternatives. I am afraid it would complicate things quite considerably: indeed, it would require either some explicit structures like 'choice', 'alternative', etc, or an extra term with a restricted vocabulary that describes what the multiple targets are there for. All these things are of course possible, I am not sure they would hit the 80/20 mark...

(As an aside, if we want to continue this discussion, we should spawn a separate issue. This is a different topic to the current one.)

**Comment by paolociccarese** [comment]
Hi @iherman, 
I am ok with having a separate discussion, however I am not sure this is a different topic. 

We can ask to use multiple targets for targeting translations as you stated above, in fact that is consistent with the current definition of multiple targets ("Body is considered to be equally related to each Target individually"). If we annotate different content formats, that is addressed by having different States. 

So, I would say that allowing multiple selectors would serve the purpose of annotating multiple targets as a single unit. Which to me seems a pretty straightforward interpretation as well.

**Comment by iherman** [comment]
> On 4 May 2016, at 15:52, Paolo Ciccarese notifications@github.com wrote:
> 
> Hi @iherman, 
> I am ok with having a separate discussion, however I am not sure this is a different topic.
> 
> We can ask to use multiple targets for targeting translations as you stated above, in fact that is consistent with the current definition of multiple targets ("Body is considered to be equally related to each Target individually"). If we annotate different content formats, that is addressed by having different States.
> 
> So, I would say that allowing multiple selectors would serve the purpose of annotating multiple targets as a single unit. Which to me seems a pretty straightforward interpretation as well.

I am not sure we are in agreement here, but I may misunderstand you.

In my understanding, the correct solution for encoding your example, ie, the reference to Paolo and Ciccarese, is to use multiple targets in the same annotation. One target is for 'Paolo' and the other is for 'Ciccarese'. Each of those targets may use several selectors, but each of those selectors  should address 'Paolo' (resp. 'Ciccarese') in some way or another.

Conceptually, at least in my mind, having several targets in the same annotation is like adding an annotation on the 'combination' (without specifying what 'combination' means) of the targets. That is the reason of keeping them in the same annotation.

Is this what you meant?

**Comment by pciccarese** [comment]
I guess it goes down to the interpretation of the  definition of 'multiple targets': "Body is considered to be equally related to each Target individually". 

The 'individually' to me means that using multiple targets does not convey the fact that those targets have to be seen as a single unit and that the annotation is targeting all of them at the same time (i.e. if one is missing the annotation is not meaningful anymore). 

I am open to consider alternative interpretations though :)

**Comment by tcole3** [comment]
It is deja vu all over again.

The model and vocabulary documents define the class oa:Choice (json key Choice) as "a subClass of as:OrderedCollection that conveys to a consuming application that it should select one of the resources in the as:items list to use, rather than all of them."  

In closing issues #92, #145, etc. I thought we had concluded that as:OrderedCollection (or a subclass of it) would be appropriate to replace both oa:List and oa:Composite (from previous iterations of the model). This would provide a means to differentiate target as array, 'Body is considered to be equally related to each Target individually,' from Paolo's use case of multiple resources grouped together as a collection to serve as annotation target. In closing #92, I do not recall agreeing to drop this use case altogether (but this could be faulty memory - pointer?). 

Since then we have defined the json key 'AnnotationCollection' to be the alias for 'as:OrderedCollection' which means we don't have a key that we can use for the equivalent of the old oa:Composite / oa:List classes when it comes to grouping resources used collectively as a target or body.  Personally I still see a need a json key for this, and as mentioned I thought this is what we agreed to collectively in closing the various multiplicity issues previously discussed. 

I appreciate that some folks will not will not be interested in the distinction between a json array and any 'Collection' class or subclass, but we've already found it necessary for Choice and for sets of Annotations, so why not for Collections formerly known as Composite / List?  

BTW, I note the old oa:List class is still to be found in Appendix A of our latest model Working Draft (https://www.w3.org/TR/annotation-model/#complete-example). I assume this is just an editorial oversight, but I choose to take it as a sign that we need something like oa:List (json key List) as another sub-class of as:OrderedCollection to handle the use case Paolo is raising.  Personally I find List a straightforward complement to Choice (no more nor less complex), it is consistent with at least what some of us understood when issues #92, #145, etc. were closed, and I do not agree that Paolo's use case fails by the 80/20 rule.

**Comment by jjett** [comment]
+1 to what Tim is saying. 

In the end it doesn't look like we actually managed to settle the composite/choice (or as I prefer to look at them -- the logical AND/OR) entities as much as we had hoped when we discussed them way back when (in was it January?).

**Comment by iherman** [comment]
Hm. There is a bit of a mess indeed. As far as I can see (and looking at targets only) we have
1. the possibility to define multiple targets, which means "each Body is considered to be equally related to each Target individually, rather than to the set of all of the Targets."
2. the possibility to define a Choice, which "A Choice has an ordered list of resources from which an application should select only one to process or display. The order is given from the most preferable to least preferable, according to the Annotation's creator or publisher. " (which is a suitable terminology for a body, but should be different for a target; I believe this is, essentially, the OR feature of @jjett )
3. @tcole3 refers to "AnnotationCollection", but I am not sure that is relevant; that is a collection of full blown annotations, and not related to targets and bodies

Ie, we need, again to use @jjett's terminology, an AND. Can we do it similarly to "Choice", calling it "Combination" or something similar? That would also cover @paolociccarese's use case. We can add this to the model easily...

(However, this is _irrelevant_ to the multiple selector issue! Can we really move this thread to a separate issue, unless we get to a quick solution?)

**Comment by jjett** [comment]
We had an AND similar to "Choice", it was named "Composite". Closed issue #92 may be relevant. 

Regarding selectors, closed issue #93 seems highly pertinent. It seems to me that the case for logical AND vis-a-vis multiple selectors looks suspiciously like the IIIF's Image API Selector which simultaneously specifies some 2-space coordinates and a rotational position on a (fixed) axis. Having a construct that allows folks to aggregate selectors in a similar fashion would allow them to remark on how the selectors are expected to behave without going the extra distance of actually specifying a particular API (as IIIF did).

I should mention that these logical structures are very closely linked to expected behaviors that are neither list-like nor set-like, despite what our serialization tools might indicate.

**Comment by iherman** [comment]
Discussed at F2F, 17.05.16

RESOLUTION:  Extend the Choice structure by adding back Composite, List, plus a new one Individuals to explicitly state each is independent. The entire section to be marked At Risk

See http://www.w3.org/2016/05/17-annotation-irc#T14-30-01-1

**Comment by azaroth42** [comment]
Note that for selectors, we will allow multiple selectors on a single specific resource, with the semantics of choose one of these, and that they should identify the same region with some, unspecified, degree of fidelity.

**Comment by dirkroorda** [comment]
That's a pity. @paolociccarese makes a perfectly reasonable case.

Suppose you have a series of annotations members of a set X of things. That series is effectively a mapping from X to annotated values.

It is perfectly reasonable that you also want to be able to map pairs of things. That would be a mapping of X x X (the cartesian product of X) to annotated values.

A generic example is when you want to annotate pairs of things by a measure of how similar they are.

In general, you cannot reduce a function from X x X to functions from X. 

By choosing the interpretation as "each Body is considered to be equally related to each Target individually, rather than to the set of all of the Targets" power is lost.

If the other choice had been made, nothing had been lost.

Because of this choice webannotations are crippled for use cases where you want to annotate relationships between resources.

The question is, which other vocabulary will come to the rescue when you want to that?
